jax/experimental/jax2tf/jax_export.py

```diff
@@ -105,103 +116,198 @@ def vjp(self) -> "Exported":
-  if "global_out_avals" in lowered.compile_args:
-    # This is currently the case for pjit
-    out_avals = lowered.compile_args["global_out_avals"]
-  elif "shards" in lowered.compile_args:  # for PmapComputation
-    out_avals = lowered.compile_args["shards"].out_sharded_avals
-  else:
-    out_avals = lowered.compile_args["out_avals"]
-
-  # Log and then check the module.
-  if logging.vlog_is_on(3):
-    mlir_module_text = mlir.module_to_string(mlir_module)
-    logmsg = f"version={xla_call_module_version} lowering_platform={lowering_platform}"
-    logging.vlog(3, "Lowered JAX module: %s\n%s", logmsg, mlir_module_text)
-
-  _check_module(mlir_module,
-                allow_non_replicated_sharding=allow_non_replicated_sharding,
-                allow_all_custom_calls=not strict_checks)
-
-  return Exported(
-      in_avals=args_avals,
-      out_avals=out_avals,
-      in_shardings=lowered.compile_args["in_shardings"],
-      out_shardings=lowered.compile_args["out_shardings"],
-      lowering_platform=lowering_platform or _default_jax_backend(),
-      strict_checks=strict_checks,
-      mlir_module_serialized=mlir_module_serialized,
-      module_kept_var_idx=module_kept_var_idx,
-      xla_call_module_version=xla_call_module_version,
-      _get_vjp=lambda exported: _export_native_vjp(wrapped_fun_jax, exported))
+    mlir_module = lowering.stablehlo()
+
+    args_avals_flat, _ = tree_util.tree_flatten(lowered.in_avals)
+    if "kept_var_idx" in lowering.compile_args:
+      module_kept_var_idx = tuple(sorted(lowering.compile_args["kept_var_idx"]))
+    else:
+      # For pmap
+      module_kept_var_idx = tuple(range(len(args_avals_flat)))
+
+    if not all(core.is_constant_shape(a.shape) for a in args_avals_flat):
+      # All arguments are kept if we have dimension variables.
+      assert len(module_kept_var_idx) == len(args_avals_flat)
+      mlir_module = _add_dim_arg_computation(mlir_module, args_avals_flat)
+
+    xla_call_module_version = 4
+    mlir_str = mlir.module_to_bytecode(mlir_module)
+    target_version = stablehlo.get_earliest_forward_compatible_version()
+    mlir_module_serialized = xla_client._xla.mlir.serialize_portable_artifact(
+        mlir_str, target_version)
+
+    # Figure out the result types and shapes
+    if "global_out_avals" in lowering.compile_args:
+      # This is currently the case for pjit
+      out_avals_flat = lowering.compile_args["global_out_avals"]
+    elif "shards" in lowering.compile_args:  # for PmapComputation
+      out_avals_flat = lowering.compile_args["shards"].out_sharded_avals
+    else:
+      out_avals_flat = lowered.compile_args["out_avals"]
+
+    # Log and then check the module.
+    if logging.vlog_is_on(3):
+      mlir_module_text = mlir.module_to_string(mlir_module)
+      logmsg = f"version={xla_call_module_version} lowering_platform={lowering_platform}"
+      logging.vlog(3, "Lowered JAX module: %s\n%s", logmsg, mlir_module_text)
```

Comment:
Not related to this PR, but sometimes the MLIR module can be too large to print. Maybe consider dumping it to a file https://github.com/google/jax/blob/main/jax/_src/dispatch.py#L481-L482?

Reply:
Good idea, I will do this in a future change